src/legacy/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable_factory.tsx

```diff
@@ -0,0 +1,191 @@
+            return 'visualizeApp';
+          }
+          return (
+            this.visTypes.byName[JSON.parse(savedObject.attributes.visState).type].icon ||
+            'visualizeApp'
+          );
+        },
+        getTooltipForSavedObject: savedObject => {
+          if (!this.visTypes) {
+            return '';
+          }
+          return `${savedObject.attributes.title} (${this.visTypes.byName[JSON.parse(savedObject.attributes.visState).type].title})`;
+        },
+        showSavedObject: savedObject => {
+          if (!this.visTypes) {
+            return false;
+          }
+          const typeName: string = JSON.parse(savedObject.attributes.visState).type;
+          const visType = this.visTypes.byName[typeName];
+          if (!visType) {
+            return false;
+          }
+          if (chrome.getUiSettingsClient().get('visualize:enableLabs')) {
+            return true;
+          }
+          return visType.stage !== 'experimental';
+        },
+      },
+    });
+    this.initializeVisTypes();
+  }
+
+  public isEditable() {
+    return capabilities.get().visualize.save as boolean;
+  }
+
+  public getDisplayName() {
+    return i18n.translate('kbn.embeddable.visualizations.displayName', {
+      defaultMessage: 'visualization',
+    });
+  }
+
+  public async initializeVisTypes() {
+    const $injector = await chrome.dangerouslyGetActiveInjector();
+    const Private = $injector.get<IPrivate>('Private');
+    this.visTypes = Private(VisTypesRegistryProvider);
+  }
+
+  public async createFromSavedObject(
+    savedObjectId: string,
+    input: Partial<VisualizeInput> & { id: string },
+    parent?: Container
+  ): Promise<VisualizeEmbeddable | ErrorEmbeddable | DisabledLabEmbeddable> {
+    const $injector = await chrome.dangerouslyGetActiveInjector();
+    const config = $injector.get<Legacy.KibanaConfig>('config');
+    const savedVisualizations = $injector.get<SavedVisualizations>('savedVisualizations');
+
+    try {
+      const visId = savedObjectId;
+
+      const editUrl = chrome.addBasePath(`/app/kibana${savedVisualizations.urlFor(visId)}`);
+      const loader = await getVisualizeLoader();
+      const savedObject = await savedVisualizations.get(visId);
+      const isLabsEnabled = config.get<boolean>('visualize:enableLabs');
+
+      if (!isLabsEnabled && savedObject.vis.type.stage === 'experimental') {
+        return new DisabledLabEmbeddable(savedObject.title, input);
+      }
+
+      const indexPattern = await getIndexPattern(savedObject);
+      const indexPatterns = indexPattern ? [indexPattern] : [];
+      return new VisualizeEmbeddable(
+        {
+          savedVisualization: savedObject,
+          loader,
+          indexPatterns,
+          editUrl,
+          editable: this.isEditable(),
+        },
+        input,
+        parent
+      );
+    } catch (e) {
+      console.error(e); // eslint-disable-line no-console
+      return new ErrorEmbeddable(e, input, parent);
```

Comment:
Why returning instead of throwing?

Reply:
I don't want an error to break an entire dashboard so it's better to return an error embeddable that shows the user the issue.

I suppose I could also have factories throw errors and make the container handle catching and converting into an error embeddable. Though it's kind of nice that the container can assume it will get an embeddable of some sort.

One common error here is if there is no saved object for that id.